File: main/java/mod/iceandshadow3/compat/Vec3.java
package mod.iceandshadow3.compat;

import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.ChunkPos;

/**
 * An immutable fixed-precision 3d vector class.
 */
public class Vec3 implements Cloneable, Comparable<Vec3> {
	protected static final long SUB_BITS = 16;
	protected static final long CHUNK_BITS = 4;
	
	protected static final long SUB_MULT = 1 << SUB_BITS;
	protected static final long CHUNK_MULT = 1 << CHUNK_BITS;
	protected static final long SUB_MASK = SUB_MULT-1;
	protected static final long CHUNK_MASK = CHUNK_MULT-1;
	protected static long toLong(double value) {
		long whole = (long)value;
		return (whole << SUB_BITS) + (long)((value - whole) * SUB_MULT);
	}
	protected static long toBlockCoord(long value) {
		return value >> SUB_BITS;
	}
	protected static int toSubCoord(long value) {
		return (int)(value & SUB_MASK);
	}
	protected static double toDouble(long value) {
		return toBlockCoord(value) + ((double)toSubCoord(value))/SUB_MULT;
	}
	
	// WARNING: Do not make these final. However, also do not modify them.
	// For in-place modifying operations, see Vec3M.
	protected long x, z;
	protected int y;
	
	/// Unit vectors.
	public static final Vec3
		U = new Vec3(0, 1, 0),
		D = new Vec3(0, -1, 0),
		S = new Vec3(0, 0, 1),
		N = new Vec3(0, 0, -1),
		E = new Vec3(1, 0, 0),
		W = new Vec3(-1, 0, 0);
		
	public long xBlock() {return toBlockCoord(x);}
	public int yBlock() {return (int)toBlockCoord(y);}
	public long zBlock() {return toBlockCoord(z);}
	public int xChunk() {return (int)(toBlockCoord(x) >> CHUNK_BITS);}
	public int zChunk() {return (int)(toBlockCoord(z) >> CHUNK_BITS);}
	public byte xSubChunk() {return (byte)(toBlockCoord(x) & CHUNK_MASK);}
	public byte zSubChunk() {return (byte)(toBlockCoord(z) & CHUNK_MASK);}
	public int xSubBlock() {return toSubCoord(x);}
	public int ySubBlock() {return toSubCoord(y);}
	public int zSubBlock() {return toSubCoord(z);}
	public double xDouble() {return toDouble(x);}
	public double yDouble() {return toDouble(y);}
	public double zDouble() {return toDouble(z);}
	
	public Vec3(int x, int y, int z) {
		this(x, y, z, true);
	}
	protected Vec3(long x, int y, long z, boolean needsshift) {
		long shift = needsshift ? SUB_BITS : 0;
		this.x = x << shift;
		this.y = y << shift;
		this.z = z << shift;
	}
	/// Returns a mutable copy of this vector.
	Vec3M copy() {
		return new Vec3M(this);
	}
	
	public double mag() {
		return Math.sqrt(x*x + (long)y*y + z*z);
	}
	public double magH() {
		return Math.sqrt(x*x + z*z);
	}
	public double dot(Vec3 b) {
		return toDouble(x*b.x + (long)y*b.y + z*b.z);
	}
	
	public BlockPos toBlockPos() {
		return new BlockPos(xBlock(), yBlock(), zBlock());
	}
	
	@Override
	public boolean equals(Object o) {
		if(o instanceof Vec3) {
			Vec3 b = (Vec3)o;
			return y==b.y && x==b.x && z==b.z;
		}
		return false;
	}
	@Override
	public String toString() {
		return "<"+xDouble()+','+yDouble()+','+zDouble()+'>';
	}
	@Override
	public int compareTo(Vec3 b) {
		return Double.compare(mag(), b.mag());
	}
	
}

File: main/scala/mod/iceandshadow3/compat/Vec3.scala
package mod.iceandshadow3.compat

import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.ChunkPos
import net.minecraft.entity.Entity
import net.minecraft.util.math.BlockPos

import Vec3._

object Vec3 {
	protected[compat] val SUB_BITS: Long = 16
	protected val CHUNK_BITS: Long = 4
	protected val SUB_MULT: Long = 1 << SUB_BITS
	protected val CHUNK_MULT: Long = 1 << CHUNK_BITS
	protected val SUB_MASK: Long = SUB_MULT - 1
	protected val CHUNK_MASK: Long = CHUNK_MULT - 1

	protected[compat] def toLong(value: Double): Long = {
		val whole: Long = value.toLong
		(whole << SUB_BITS) + ((value - whole) * SUB_MULT).toLong
	}

	protected def toBlockCoord(value: Long): Long = value >> SUB_BITS
	protected def toSubCoord(value: Long): Int = (value & SUB_MASK).toInt
	protected[compat] def toDouble(value: Long): Double =
		toBlockCoord(value) + toSubCoord(value).toDouble / SUB_MULT

	// Unit vectors.
	val U: Vec3 = new Vec3(0, 1, 0)
	val D: Vec3 = new Vec3(0, -1, 0)
	val S: Vec3 = new Vec3(0, 0, 1)
	val N: Vec3 = new Vec3(0, 0, -1)
	val E: Vec3 = new Vec3(1, 0, 0)
	val W: Vec3 = new Vec3(-1, 0, 0)
}

/**
* An immutable fixed-precision 3d vector class.
*/
class Vec3 protected (
	protected[compat] var x: Long,
	protected[compat] var y: Int,
	protected[compat] var z: Long
) extends Comparable[Vec3] with Cloneable {
	protected def this(x: Long, y: Int, z: Long, shift: Long) = {
		this(x << shift, y << shift, z << shift)
	}
	def this(x: Int, y: Int, z: Int) = this(x, y, z, SUB_BITS)
	def this(p: BlockPos) = this(p.getX(), p.getY(), p.getZ(), SUB_BITS)
	def this(e: Entity) = this(toLong(e.posX), toLong(e.posY).toInt, toLong(e.posZ), 0)

	def xBlock(): Long = toBlockCoord(x)
	def yBlock(): Int = toBlockCoord(y).toInt
	def zBlock(): Long = toBlockCoord(z)
	def xChunk(): Int = (toBlockCoord(x) >> CHUNK_BITS).toInt
	def zChunk(): Int = (toBlockCoord(z) >> CHUNK_BITS).toInt
	def xSubChunk(): Byte = (toBlockCoord(x) & CHUNK_MASK).toByte
	def zSubChunk(): Byte = (toBlockCoord(z) & CHUNK_MASK).toByte
	def xSubBlock(): Int = toSubCoord(x)
	def ySubBlock(): Int = toSubCoord(y)
	def zSubBlock(): Int = toSubCoord(z)
	def xDouble(): Double = toDouble(x)
	def yDouble(): Double = toDouble(y)
	def zDouble(): Double = toDouble(z)

/// Returns a mutable copy of this vector.
	def copy(): Vec3M = new Vec3M(x, y, z)
	def mag(): Double = Math.sqrt(x * x + y.toLong * y + z * z)
	def magH(): Double = Math.sqrt(x * x + z * z)

	def dot(b: Vec3): Double = toDouble(x * b.x + y.toLong * b.y + z * b.z)
	def toBlockPos(): BlockPos = new BlockPos(xBlock(), yBlock(), zBlock())

	override def equals(o: Any): Boolean = {
		if (o.isInstanceOf[Vec3]) {
			val b: Vec3 = o.asInstanceOf[Vec3]
			y == b.y && x == b.x && z == b.z
		} else false
	}

	override def toString(): String = s"<$xDouble,$yDouble,$zDouble()>"

	override def compareTo(b: Vec3): Int = java.lang.Double.compare(mag(), b.mag())
}

